python/rikai/spark/sql/codegen/tensorflow.py

```diff
@@ -0,0 +1,100 @@
+#  Copyright 2022 Rikai Authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Iterator
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from pyspark.serializers import CloudPickleSerializer
+from pyspark.sql.functions import pandas_udf
+from pyspark.sql.types import BinaryType
+
+from rikai.spark.sql.codegen.base import ModelSpec
+from rikai.torch.pandas import PandasDataset
+from rikai.types import Image
+
+DEFAULT_BATCH_SIZE = 4
+
+
+__all__ = ["generate_udf", "load_model_from_uri"]
+
+_pickler = CloudPickleSerializer()
+
+
+def infer_output_signature(blob: bytes):
+    row = _pickler.loads(blob)
+
+    if isinstance(row, Image):
+        image_arr = row.to_numpy()
+        if len(image_arr.shape) == 2:
+            return tf.TensorSpec(shape=(None, None), dtype=tf.uint8)
+        elif len(image_arr.shape) == 3:
+            return tf.TensorSpec(
+                shape=(
+                    None,
+                    None,
+                    image_arr.shape[2],
+                ),
+                dtype=tf.uint8,
+            )
+        else:
+            raise ValueError(
+                f"Image should only have 2d or 3d shape, got {image_arr.shape}"
```

Comment:
do we want to support text types through this interface?

Reply:
do you mean `text` as a new type? If so, yes. So i was thinking to put a `classmethod` to `ToNumpy` mixins, and let it return the shape of the tensor / ndarray:

```
class ToNumpy(abc.ABC):
    @abstractmethod
    def shape(self):
         pass
```

But dont have concrete idea how to do it yet. The main blocker is still the uncertainty of the shape defined in `NDArray`.